crates/ruff_cli/src/cache.rs

```diff
@@ -1,171 +1,235 @@
+            msg.file.source_text().to_owned()
+        } else {
+            String::new() // No messages, no need to keep the source!
+        };
 
-        s.serialize_field("kind", &kind)?;
-        s.serialize_field("range", &range)?;
-        s.serialize_field("fix", &fix)?;
-        s.serialize_field("file_id", &self.file_id)?;
-        s.serialize_field("noqa_row", &noqa_row)?;
+        let messages = messages
+            .iter()
+            .map(|msg| CacheMessage {
+                kind: msg.kind.clone(),
+                range: msg.range,
+                fix: msg.fix.clone(),
+                noqa_offset: msg.noqa_offset,
+            })
+            .collect();
+
+        FileCache {
+            last_modified,
+            imports: imports.clone(),
+            messages,
+            source,
+        }
+    }
 
-        s.end()
+    /// Convert the file cache into `Diagnostics`, using `path` as file name.
+    pub(crate) fn into_diagnostics(self, path: &Path) -> Diagnostics {
+        let messages = if self.messages.is_empty() {
+            Vec::new()
+        } else {
+            let file = SourceFileBuilder::new(path.to_string_lossy(), self.source).finish();
+            self.messages
+                .into_iter()
+                .map(|msg| Message {
+                    kind: msg.kind,
+                    range: msg.range,
+                    fix: msg.fix,
+                    file: file.clone(),
+                    noqa_offset: msg.noqa_offset,
+                })
+                .collect()
+        };
+        Diagnostics::new(messages, self.imports)
     }
 }
 
-#[derive(Deserialize)]
-struct MessageHeader {
+/// On disk representation of a diagnostic message.
+#[derive(Clone, Deserialize, Debug, Serialize)]
+struct CacheMessage {
     kind: DiagnosticKind,
+    /// Range into the message's [`FileCache::source`].
     range: TextRange,
     fix: Option<Fix>,
-    file_id: usize,
-    noqa_row: TextSize,
+    noqa_offset: TextSize,
 }
 
-#[derive(Deserialize)]
-struct CheckResult {
-    imports: ImportMap,
-    messages: Vec<MessageHeader>,
-    files: FilesBuf,
-}
-
-fn content_dir() -> &'static Path {
-    Path::new("content")
-}
-
-fn cache_key(
-    path: &Path,
-    package: Option<&Path>,
-    metadata: &fs::Metadata,
-    settings: &Settings,
-) -> u64 {
+/// Returns a hash key based on the `package_root`, `settings` and the crate
+/// version.
+fn cache_key(package_root: &Path, settings: &Settings) -> u64 {
     let mut hasher = CacheKeyHasher::new();
-    CARGO_PKG_VERSION.cache_key(&mut hasher);
-    path.absolutize().unwrap().cache_key(&mut hasher);
-    package
-        .as_ref()
-        .map(|path| path.absolutize().unwrap())
-        .cache_key(&mut hasher);
-    FileTime::from_last_modification_time(metadata).cache_key(&mut hasher);
-    #[cfg(unix)]
-    metadata.permissions().mode().cache_key(&mut hasher);
+    env!("CARGO_PKG_VERSION").cache_key(&mut hasher);
```

Comment:
You inherited this code and it's fine as currently the difference between cache and no-cache still isn't so big, but i wonder if it makes sense to switch to something else, is there e.g. a library that could hash our type layout or something?

Reply:
that's totally fair!